gobblin-api/src/main/java/gobblin/util/ClassAliasResolver.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (C) 2014-2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+package gobblin.util;
+
+import java.util.Map;
+
+import lombok.extern.slf4j.Slf4j;
+
+import org.reflections.Reflections;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+
+import gobblin.annotation.Alias;
+
+
+/**
+ * A thread safe singleton that lazily scans through all classes in the classpath annotated with {@link Alias} and caches
+ * the alias to class mapping. Applications can call {@link #resolve(String)} to resolve a class's alias to its
+ * cannonical name.
+ *
+ * <br>
+ * <b>Note : If same alias is used for multiple classes in the classpath, the first mapping in classpath scan holds good.
+ * Subsequent ones will be logged and ignored.
+ * </b>
+ */
+@Slf4j
+public class ClassAliasResolver {
+
+  private static class ClassAliasResolverHolder {
+    private static final ClassAliasResolver INSTANCE = new ClassAliasResolver();
+  }
+
+  Map<String, Class<?>> aliasToClassCache;
```

Comment:
Assuming the aliases are globally unique seems a bit dangerous. For example, I may have "hourly" alias for date-partitioned dataset finder and for a copy policy. Previous design where the aliased class had to be a subclass of a root interface seemed more sound. The factory class that is used to instantiate implementations of that interface can have its own (non-static) copy of the ClassAliasResolver.